src/crypto/mod.rs

```diff
@@ -14,28 +15,359 @@ use serde_cbor::Value;
 use std::convert::TryFrom;
 use std::fmt;
 
-cfg_if::cfg_if! {
-    if #[cfg(feature = "crypto_ring")] {
-        #[path = "ring.rs"]
-        pub mod imp;
-    } else if #[cfg(feature = "crypto_openssl")] {
-        #[path = "openssl.rs"]
-        pub mod imp;
-    } else if #[cfg(feature = "crypto_dummy")] {
-        #[path = "dummy.rs"]
-        pub mod imp;
-    } else {
-        #[path = "nss.rs"]
-        pub mod imp;
+#[cfg(feature = "crypto_nss")]
+mod nss;
+#[cfg(feature = "crypto_nss")]
+use nss as backend;
+
+#[cfg(feature = "crypto_openssl")]
+mod openssl;
+#[cfg(feature = "crypto_openssl")]
+use self::openssl as backend;
+
+#[cfg(feature = "crypto_dummy")]
+mod dummy;
+#[cfg(feature = "crypto_dummy")]
+use dummy as backend;
+
+use backend::{
+    decrypt_aes_256_cbc_no_pad, ecdhe_p256_raw, encrypt_aes_256_cbc_no_pad, hmac_sha256,
+    random_bytes, sha256,
+};
+
+// Object identifiers in DER tag-length-value form
+const DER_OID_EC_PUBLIC_KEY_BYTES: &[u8] = &[
+    0x06, 0x07,
+    /* {iso(1) member-body(2) us(840) ansi-x962(10045) keyType(2) ecPublicKey(1)} */
+    0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01,
+];
+const DER_OID_P256_BYTES: &[u8] = &[
+    0x06, 0x08,
+    /* {iso(1) member-body(2) us(840) ansi-x962(10045) curves(3) prime(1) prime256v1(7)} */
+    0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07,
+];
+
+pub struct PinUvAuthProtocol(Box<dyn PinProtocolImpl + Send + Sync>);
```

Comment:
Unfortunately not, trait objects are limited to one object-safe base trait and any number of "auto-traits". And Clone isn't an auto-trait.

Reply:
I don't suppose it can be something like `pub struct PinUvAuthProtocol(Box<dyn T + Send + Sync>) where T : PinProtocolImpl, Clone`? Anyway, not a big deal.